src/Microsoft.ManagedZlib/src/IHuffmanTree.cs

```diff
@@ -0,0 +1,325 @@
+        {
+            // bits for Literal/Length tree table
+            _tableBits = 9;
+        }
+        else
+        {
+            // bits for distance tree table and code length tree table
+            _tableBits = 7;
+        }
+        _tableMask = (1 << _tableBits) - 1;
+
+        _table = new short[1 << _tableBits];
+
+        // I need to find proof that left and right array will always be
+        // enough. I think they are.
+        _left = new short[2 * _codeLengthArray.Length];
+        _right = new short[2 * _codeLengthArray.Length];
+
+        CreateTable();
+    }
+
+    // Generate the array contains huffman codes lengths for static huffman tree.
+    // The data is in RFC 1951.
+    private static byte[] GetStaticLiteralTreeLength()
+    {
+        byte[] literalTreeLength = new byte[MaxLiteralTreeElements];
+
+        literalTreeLength.AsSpan(0, 144).Fill(8);
+        literalTreeLength.AsSpan(144, 112).Fill(9);
+        literalTreeLength.AsSpan(256, 24).Fill(7);
+        literalTreeLength.AsSpan(280, 8).Fill(8);
+        return literalTreeLength;
+    }
+
+    private static byte[] GetStaticDistanceTreeLength()
+    {
+        byte[] staticDistanceTreeLength = new byte[MaxDistTreeElements];
+        Array.Fill(staticDistanceTreeLength, (byte)5);
+        return staticDistanceTreeLength;
+    }
+
+    // Reverse 'length' of the bits in code
+    private static uint BitReverse(uint code, int length)
+    {
+        uint new_code = 0;
+
+        Debug.Assert(length > 0 && length <= 16, "Invalid len");
+        do
+        {
+            new_code |= (code & 1);
+            new_code <<= 1;
+            code >>= 1;
+        } while (--length > 0);
+
+        return new_code >> 1;
+    }
+
+    // Calculate the huffman code for each character based on the code length for each character.
+    // This algorithm is described in standard RFC 1951
+    private uint[] CalculateHuffmanCode()
+    {
+        Span<uint> bitLengthCount = stackalloc uint[17];
+        bitLengthCount.Clear();
+        foreach (int codeLength in _codeLengthArray)
+        {
+            bitLengthCount[codeLength]++;
+        }
+        bitLengthCount[0] = 0;  // clear count for length 0
+
+        Span<uint> nextCode = stackalloc uint[17];
+        nextCode.Clear();
+        uint tempCode = 0;
+        for (int bits = 1; bits <= 16; bits++)
+        {
+            tempCode = (tempCode + bitLengthCount[bits - 1]) << 1;
+            nextCode[bits] = tempCode;
+        }
+
+        uint[] code = new uint[MaxLiteralTreeElements];
+        for (int i = 0; i < _codeLengthArray.Length; i++)
+        {
+            int len = _codeLengthArray[i];
+
+            if (len > 0)
+            {
+                code[i] = BitReverse(nextCode[len], len);
+                nextCode[len]++;
+            }
+        }
+        return code;
+    }
+
+    private void CreateTable()
+    {
+        uint[] codeArray = CalculateHuffmanCode();
+#if DEBUG
```

Comment:
Was this a translation from native code? Do we still need this?

Reply:
This is used for decoding dynamic block headers (format of compression with Dynamic Huffman codes) in Inflate.